File: Algo/Leetcode/501ModeInBST.cpp
// store the last value and counts. Update by dfs.
class Solution {
public:
    int max_count = 0;
    vector<int> res;
    int count = 0;
    TreeNode* pre = NULL;
    int last_val = 0;
    vector<int> findMode(TreeNode* root) {
        if(!root) return res;
        dfs(root);
        return res;
    }

    void dfs(TreeNode* cur){
        if(!cur) return;

        dfs(cur->left);
        if(!pre){
            last_val = cur->val;
            pre = cur;
            max_count = 1;
            count = 1;
            res.push_back(cur->val);
        } else{
            if(cur->val == last_val) count += 1;
            else count = 1;
            last_val = cur->val;
            if(count == max_count) res.push_back(cur->val);
            else if(count > max_count){
                while(!res.empty()) res.pop_back();
                res.push_back(cur->val);
            }
            max_count = max(max_count, count);
        }
        pre = cur;
        dfs(cur->right);
    }
};

File: Algo/Leetcode/501ModeInBST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    def __init__(self):
        self.prev = None
        self.mode = []
        self.count = 0
        self.max_count = 0

    def findMode(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root is None:
            return

        self.dfs(root)

        return self.mode

    def dfs(self, root):
        if root is None:
            return

        self.dfs(root.left)

        val = root.val
        if self.prev is None:
            self.prev = val
            self.count = 1
        else:
            if val == self.prev:
                self.count += 1
            else:
                self.count = 1
                self.prev = val
        # update
        if self.count > self.max_count:
            self.max_count = self.count
            self.mode = [val]
        elif self.count == self.max_count:
            self.mode.append(val)

        self.dfs(root.right)
